content/post/calling-c-functions-from-bigquery.md

```diff
@@ -0,0 +1,208 @@
+
+This is using Legacy SQL, we'll see later why Standard SQL would make it much harder for us.
+
+{{% code "/post/c-on-bigquery/udf.js" javascript %}}
+{{% code "/post/c-on-bigquery/udf.sql" sql %}}
+
+This is cool, but can we do it over multiple values?
+
+{{% code "/post/c-on-bigquery/udf-mult.js" javascript %}}
+{{% code "/post/c-on-bigquery/udf-mult.sql" sql %}}
+
+Great, so we can basically write any Javascript and have it executed as part of our SQL queries.
+Unfortunately for us, the library that I would like to call (libuast) is actually written in C!
+
+How to solve this conundrum? Enter wasm.
+
+## Web Assembly
+
+Web Assembly (a.k.a wasm) is a relatively new technology that proposes an assembly language for the web.
+Javascript can be compiled to this assembly, but many other languages, such as C, Rust, and hopefully soon Go,
+can also be compiled to highly efficient web assembly code.
+
+Could we maybe compile our C libraries to wasm and then load them from Javascript?
+Well, it's technically possible and that's how much I need to give it a try.
+
+### Compiling C
+
+Compiling C programs is an incredibly interesting topic.
+Lots can be said, but I will keep it to the basics and talk about what objects are,
+how to build them, and how to link them either statically or dynamically.
+
+If you already know all of this feel free to skip this part and move directly to how
+to compile C to wasm.
+
+{{% code "/post/c-on-bigquery/sum.c" c %}}
+
+We can compile this piece file into a library.
+
+```bash
+> clang -c sum.c
+> file sum.o
+sum.o: Mach-O 64-bit object x86_64
+```
+
+In order to compile I wrote a `main` function that predeclares `sum` and calls it.
+
+{{% code "/post/c-on-bigquery/main.c" javascript %}}
+
+When we compile and execute this program the output is the answer to everything.
+
+```bash
+> clang -o main main.c sum.o
+> ./main
+42
+```
+
+When we compile it in this way we are letting the C linker unify the `sum` declared in
+`main.c` and realize that it is the same as the one defined in `sum.c`.
+This is static linking, but we could also do something similar with dynamic linking,
+letting the binary find and identify `sum` at runtime.
+
+{{% code "/post/c-on-bigquery/dynamic.c" c %}}
+
+```bash
+> clang sum.o -shared -o sum.so
+> file sum.so
+sum.so: Mach-O 64-bit dynamically linked shared library x86_64
+> clang -o dynamic dynamic.c
+> ./dynamic
+```
+
+Try it yourself, what happens if you delete `sum.o` and run `main`?
+What if you delete `sum.so` and run `dynamic`?
+
+### Compiling C to wasm
+
+Now that we understand how to dynamically load and execute shared objects,
+let's talk about wasm.
+In wasm we are going to generate a shared object from `sum.c`, but one that
+rather than being destined to be read by C it is for wasm.
+
+The easiest way I've seen to use `emcc` (also known as emscripten) is with
```

Comment:
Shall we also add http://emscripten.org ?

Reply:
sure!